builder/slide_and_fold_storage.cc

```diff
@@ -224,13 +260,102 @@ class slide_and_fold : public node_mutator {
     set_result(allocate::make(op->sym, op->storage, op->elem_size, std::move(dims), body));
   }
 
+  void try_to_fold(const var& output, bool did_overlapped_fold) {
+    // Only consider loops that are inside the allocation of this output for folding.
+    // TODO: It seems like there's probably a more elegant way to do this.
+    std::optional<std::size_t> alloc_loop_level = allocation_loop_levels[output];
+    if (!alloc_loop_level) alloc_loop_level = 1;
+
+    loop_info& loop = loops.back();
+    std::optional<box_expr>& bounds = (*loop.buffer_bounds)[output];
+    if (!bounds) return;
+
+    // We don't want to use the bounds of the loop we are sliding over here.
+    auto ignore_loop_bounds = set_value_in_scope(*loop.expr_bounds, loop.sym, interval_expr::all());
+
+    expr loop_var = variable::make(loop.sym);
+
+    for (int d = 0; d < static_cast<int>(bounds->size()); ++d) {
+      interval_expr cur_bounds_d = (*bounds)[d];
+      if (!depends_on(cur_bounds_d, loop.sym).any()) {
+        // TODO: In this case, the func is entirely computed redundantly on every iteration. We should be able to
+        // just compute it once.
+        continue;
+      }
+
+      interval_expr prev_bounds_d = {
+          substitute(cur_bounds_d.min, loop.sym, loop_var - loop.step),
+          substitute(cur_bounds_d.max, loop.sym, loop_var - loop.step),
+      };
+
+      // A few things here struggle to simplify when there is a min(loop_max, x) expression involved, where x is
+      // some expression that is bounded by the loop bounds. This min simplifies away if we know that x <= loop_max,
+      // but the simplifier can't figure that out. As a hopefully temporary workaround, we can just substitute
+      // infinity for the loop max.
+      auto ignore_loop_max = [&](const expr& e) { return substitute(e, loop.bounds.max, positive_infinity()); };
+
+      interval_expr overlap = prev_bounds_d & cur_bounds_d;
+      if (prove_true(ignore_loop_max(overlap.empty()))) {
+        // The bounds of each loop iteration do not overlap. We can't re-use work between loop iterations, but we
+        // can fold the storage.
+        expr fold_factor = simplify(bounds_of(ignore_loop_max(cur_bounds_d.extent()), *loop.expr_bounds).max);
+        if (is_finite(fold_factor) && !depends_on(fold_factor, loop.sym).any()) {
+          vector_at(fold_factors[output], d) = {fold_factor, expr()};
```

Comment:
If you make this `{fold_factor, fold_factor}` (i.e. overlap = fold_factor), then I don't think you need to branch in the async code below. Just always set `fold_factor += (worker_count - 1) * overlap`, and it will do the right thing in this case too.

Reply:
Thanks, that's a nice trick!